### PR TITLE
Final bits of Drawable

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/draw/drawable.cpp [hunk 1]
```diff
@@ -101,16 +106,20 @@ bool Drawable::Is_Visible() const
 
 void Drawable::Friend_Lock_Dirty_Stuff_For_Iteration()
 {
-#ifdef GAME_DLL
-    Call_Function<void>(PICK_ADDRESS(0x004794C0, 0x007CB858));
-#endif
+    if (!s_modelLockCount && g_theGameClient != nullptr) {
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: s_modelLockCount == 0

File: src/game/client/draw/drawable.cpp [hunk 2]
```diff
@@ -133,20 +142,48 @@ const Vector3 *Drawable::Get_Selection_Color() const
 
 DrawModule **Drawable::Get_Draw_Modules()
 {
-#ifdef GAME_DLL
-    return Call_Method<DrawModule **, Drawable>(PICK_ADDRESS(0x00475F40, 0x007C87E4), this);
-#else
-    return nullptr;
-#endif
+    DrawModule **modules = (DrawModule **)Get_Modules(MODULE_DRAW);
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp:145]: Remove cast.

File: src/game/client/draw/drawable.cpp [hunk 3]
```diff
@@ -133,20 +142,48 @@ const Vector3 *Drawable::Get_Selection_Color() const
...
 }
 
 const DrawModule **Drawable::Get_Draw_Modules() const
 {
-#ifdef GAME_DLL
-    return Call_Method<const DrawModule **, const Drawable>(PICK_ADDRESS(0x00475F40, 0x007C88A2), this);
-#else
-    return nullptr;
-#endif
+    const DrawModule **modules = (const DrawModule **)Get_Modules(MODULE_DRAW);
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp:169]: Remove cast.

File: src/game/client/draw/drawable.cpp [hunk 4]
```diff
@@ -133,20 +142,48 @@ const Vector3 *Drawable::Get_Selection_Color() const
...
+
+    if (m_isModelDirty) {
+        if (s_modelLockCount <= 0) {
+            for (const DrawModule **i = modules; *i != nullptr; i++) {
+                ObjectDrawInterface *draw = ((DrawModule *)*i)->Get_Object_Draw_Interface();
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: Remove cast.

File: src/game/client/draw/drawable.cpp [hunk 5]
```diff
@@ -3289,3 +3326,482 @@ void Drawable::Color_Tint(RGBColor const *color)
...
+        }
+    }
+}
+
+void Drawable::Saturate_RGB(RGBColor &color, float factor)
+{
+    color.red *= factor;
+    color.green *= factor;
+    color.blue *= factor;
+    float mult = factor * 0.5;
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: 0.5f

File: src/game/client/draw/drawable.cpp [hunk 6]
```diff
@@ -3289,3 +3326,482 @@ void Drawable::Color_Tint(RGBColor const *color)
...
+
+void Drawable::Set_Drawable_Hidden(bool hidden)
+{
+    if (hidden != m_hidden) {
+        m_hidden = hidden;
+        Update_Hidden_Status();
+    }
+}
+
+void Drawable::Set_Effective_Opacity(float opacity1, float opacity2)
```
Comment (tomsons26) [inline src/game/client/draw/drawable.cpp:3708]: flip all >=

File: src/game/client/draw/drawable.cpp [hunk 7]
```diff
@@ -133,20 +142,48 @@ const Vector3 *Drawable::Get_Selection_Color() const
...
 {
-#ifdef GAME_DLL
-    return Call_Method<DrawModule **, Drawable>(PICK_ADDRESS(0x00475F40, 0x007C87E4), this);
-#else
-    return nullptr;
-#endif
+    DrawModule **modules = (DrawModule **)Get_Modules(MODULE_DRAW);
+
+    if (m_isModelDirty) {
+        if (s_modelLockCount <= 0) {
```
Comment (tomsons26) [inline src/game/client/draw/drawable.cpp]: flip and early return

File: src/game/client/draw/drawable.cpp [hunk 8]
```diff
@@ -133,20 +142,48 @@ const Vector3 *Drawable::Get_Selection_Color() const
...
 {
-#ifdef GAME_DLL
-    return Call_Method<const DrawModule **, const Drawable>(PICK_ADDRESS(0x00475F40, 0x007C88A2), this);
-#else
-    return nullptr;
-#endif
+    const DrawModule **modules = (const DrawModule **)Get_Modules(MODULE_DRAW);
+
+    if (m_isModelDirty) {
+        if (s_modelLockCount <= 0) {
```
Comment (tomsons26) [inline src/game/client/draw/drawable.cpp]: flip and early return

File: src/game/client/draw/drawable.cpp [hunk 9]
```diff
@@ -3289,3 +3326,482 @@ void Drawable::Color_Tint(RGBColor const *color)
...
+void Drawable::Set_Caption_Text(Utf16String const &caption)
+{
+    if (caption.Is_Empty()) {
+        Clear_Caption_Text();
+    } else {
+        Utf16String str(caption);
+        g_theLanguageFilter->Filter_Line(str);
+
+        if (m_captionText == nullptr) {
+            m_captionText = g_theDisplayStringManager->New_Display_String();
```
Comment (tomsons26) [inline src/game/client/draw/drawable.cpp:3687]: missing adjustFontSize call

File: src/game/client/draw/drawable.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp:145]: cast needs to be there.

File: src/game/client/draw/drawable.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp:169]: cast needs to be there.

File: src/game/client/draw/drawable.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: cast needs to be there.

File: src/game/client/draw/drawable.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp:3708]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp:3687]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 2 above)
Comment (xezon) [reply src/game/client/draw/drawable.cpp:145]: Hmm. static_cast then?

File: src/game/client/draw/drawable.cpp (hunk 3 above)
Comment (xezon) [reply src/game/client/draw/drawable.cpp:169]: Hmm. static_cast then?

File: src/game/client/draw/drawable.cpp (hunk 4 above)
Comment (xezon) [reply src/game/client/draw/drawable.cpp]: Hmm. static_cast then?

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**